src/client/datascience/delayedFutureExecute.ts

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+import type { JSONObject } from '@lumino/coreutils';
+import type { Kernel, KernelMessage } from '@jupyterlab/services';
+import { createDeferred } from '../common/utils/async';
+import { CancellationError } from '../common/cancellation';
+import { traceInfoIfCI } from '../common/logger';
+
+// Wraps a future so that a requestExecute on a session will wait for the previous future to finish before actually executing
+export class DelayedFutureExecute
+    implements Kernel.IShellFuture<KernelMessage.IExecuteRequestMsg, KernelMessage.IExecuteReplyMsg> {
+    private doneDeferred = createDeferred<KernelMessage.IExecuteReplyMsg>();
+    private requestFuture:
+        | Kernel.IShellFuture<KernelMessage.IExecuteRequestMsg, KernelMessage.IExecuteReplyMsg>
+        | undefined;
+    private pendingOnReply: ((msg: KernelMessage.IExecuteReplyMsg) => void | PromiseLike<void>) | undefined;
+    private pendingOnIOPub:
+        | ((msg: KernelMessage.IIOPubMessage<KernelMessage.IOPubMessageType>) => void | PromiseLike<void>)
+        | undefined;
+    private pendingOnStdin:
+        | ((msg: KernelMessage.IStdinMessage<KernelMessage.StdinMessageType>) => void | PromiseLike<void>)
+        | undefined;
+    private pendingHooks: ((
+        msg: KernelMessage.IIOPubMessage<KernelMessage.IOPubMessageType>
+    ) => boolean | PromiseLike<boolean>)[] = [];
+    private pendingInputReplies: (
+        | KernelMessage.IReplyErrorContent
+        | KernelMessage.IReplyAbortContent
+        | KernelMessage.IInputReply
+    )[] = [];
+    private disposed = false;
+    private statusChangedHandler: (_session: Kernel.IKernelConnection, status: KernelMessage.Status) => void;
+    constructor(
+        private kernelConnection: Kernel.IKernelConnection,
+        previousLink: Kernel.IShellFuture<KernelMessage.IExecuteRequestMsg, KernelMessage.IExecuteReplyMsg>,
+        private content: KernelMessage.IExecuteRequestMsg['content'],
+        private disposeOnDone?: boolean,
+        private metadata?: JSONObject
+    ) {
+        // Setup our request based on the previous link finishing
+        if (previousLink) {
```

Comment:
This is minor, but was wondering on the typing here a tiny bit. Aren't we guaranteed a previousLink at this point?

Reply:
Oh yeah that's true. I can skip the if check.